File: core/java/src/main/java/hara/lang/protocol/IStreamOffer.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Non-blocking writable stream capability. */
@HaraProtocolBinding(
    namespace = "std.protocol.istreamoffer",
    name = "IStreamOffer",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IStreamOffer {
  @HaraMethod(value = "offer", arity = 2)
  boolean offer(Object value);
}

File: core/rust/src/lang/protocol/istreamoffer.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.istreamoffer",
    name = "IStreamOffer",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IStreamOffer {
    type Value;

    #[hara_method(value = "offer", arity = 2)]
    fn offer(&mut self, value: Self::Value) -> bool;
}
